magda-gateway/src/Authenticator.ts

```diff
@@ -254,51 +269,62 @@ export default class Authenticator {
                 }
             );
         } else {
-            // For other routes: only make session & passport data available if session has already started (cookie set)
-            if (!req.cookies[DEFAULT_SESSION_COOKIE_NAME]) {
-                // --- session not started yet & no incoming session id cookie
-                // --- proceed to other non auth middlewares
-                return next();
-            } else {
-                // --- run the session middleware only first
-                return runMiddlewareList(
-                    [this.sessionMiddleware],
-                    req,
-                    res,
-                    () => {
-                        // --- check if the original incoming session id was invalid
-                        // --- here, we test session middleware's processing result
-                        // --- rather than accessing session store directly by ourself
-                        const sessionId = getSessionId(req, this.sessionSecret);
-                        if (req?.session?.id !== sessionId) {
-                            // --- a new session has been created
-                            // --- the original incoming session id must have been an invalid or expired one
-                            // --- we need to destroy this newly created empty session
-                            // --- destroy session here & no need to wait till `destroySession` complete
-                            this.destroySession(req).catch((err) => {
-                                // --- only log here if failed to delete session data from session store
-                                console.log(
-                                    `Failed to destory session: ${err}`
-                                );
-                            });
-                            this.deleteCookie(res);
-                            // --- proceed to other middleware & no need to run passport
-                            return next();
-                        } else {
-                            // --- if the session id is valid, run passport middleware
-                            return runMiddlewareList(
-                                [
-                                    this.passportMiddleware,
-                                    this.passportSessionMiddleware
-                                ],
+            // For other routes:
+            // - if valid API key headers exist, attemp to login via API key
+            // - otherwise, only make session & passport data available if session has already started (cookie set)
+            this.apiKeyMiddleware(req, res, () => {
```

Comment:
I've re-arranged the code a bit to make the purpose more clear.
The apiKeyMiddleware intend to be a standard middleware so that It can be used somewhere else if needed.
e.g. if client not happy with extra basic auth header required when basic auth is on, we can move this middleware out the authenticator and make the logic that as long as API key is correct, basic auth is not require
What's your idea on this?

Reply:
Ah yeah, fair enough